evm/src/fixed_recursive_verifier.rs

```diff
@@ -674,6 +724,88 @@ where
         }
     }
 
+    fn connect_block_proof(
+        builder: &mut CircuitBuilder<F, D>,
+        lhs: &PublicValuesTarget,
+        rhs: &PublicValuesTarget,
+    ) {
+        // Between blocks, we only connect state tries.
+        for (&limb0, limb1) in lhs
+            .trie_roots_after
+            .state_root
+            .iter()
+            .zip(rhs.trie_roots_before.state_root)
+        {
+            builder.connect(limb0, limb1);
+        }
+
+        // Connect block numbers.
+        let one = builder.one();
+        let prev_block_nb = builder.sub(rhs.block_metadata.block_number, one);
+        builder.connect(lhs.block_metadata.block_number, prev_block_nb);
```

Comment:
Should we also add a constraint of the form `(1-has_parent_block) * agg_root_proof.block_number = 0` somewhere to ensure that the genesis block has `block_number=0` (and thus to ensure there's no underflow here except for the genesis block)?

Also applies to checking that the genesis block has a predetermined state trie root.

Reply:
For the genesis block's state trie, I added a check that the state trie was empty.

Reply:
We can leave it as a TODO for now, but I think it could be useful to support non-empty genesis state tries e.g. for premines.

Reply:
It makes sense, thank you! I removed the state trie check for genesis blocks and added a TODO instead.